### PR TITLE
Migrate to Scala 3.

### PR DESCRIPTION
**Pull request by rpiaggio**
Lacking existential types, we have to pass around `S` and `B` types everywhere, but on the other hand Scala 3 can correctly infer these (and `Props`) where Scala 2 couldn't break cyclic dependencies.

For the moment syntax is not migrated (`using`, `given`, no braces, etc.).

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -101,7 +102,7 @@ lazy val root = (project in file("."))
   .aggregate(common, cats, test)
 
 lazy val commonSettings = Seq(
-  scalaVersion           := "2.13.6",
+  scalaVersion           := "3.0.2",
```
Comment (cquiroz) [inline build.sbt:105]: I think we should try to cross compile to 2.13 and 3 at least for a while?

Comment (rpiaggio) [reply build.sbt:105]: Let's just not merge it yet until we are ready to migrate everything to Scala 3. I'll change the PR to draft.

The reason being that the API changes, so we should not publish for both versions of Scala with the same library version.

In Scala 3 we have more type parameters (3) for `ReactProps` and friends which the compiler infers. In Scala 2 the compiler can't infer and we have only 1 type parameter which must be explicitly set (and the other 2 can be omitted taking advantage of existentials).

Comment (cquiroz) [reply build.sbt:105]: I see, this will ripple through the facades then. Maybe we should publish only when we need it

Comment (cquiroz) [reply build.sbt:105]: Could we add a fake type param to the scala 2 version?

**Pull request closed without merging**